### PR TITLE
Get daily saved exports off the couch

### PR DESCRIPTION
**Pull request by NoahCarnahan**
First steps at making daily saved exports compatible with the sql backend. This PR rewrites the "rebuild export" export workflow to work with the new models. At this point you can kick off a rebuild from the export list, but I still need to write the view for downloading a daily saved export and trigger the initial and periodic building of those exports.

The changes involved a lot of slightly adjusting the functionality of existing functions, while still keeping those existing functions around until the switch over is complete. To try make it clear what things are new and what things are just modifications to existing functions, I tried to put the copying steps and the modification steps into separate commits. Hopefully this makes the commit log easier to read. If it doesn't, let me know.
EDIT: Ugh, github is displaying the commits in the wrong order, despite the fact that I rebased my local changes before pushing, so things are going to look a bit jumbled up. I don't know of any way to fix it...

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 8 errors:

* corehq/apps/export/export.py, line 230 - E302 expected 2 blank lines, found 1
* corehq/apps/export/models/new.py, line 950 - E302 expected 2 blank lines, found 0
* corehq/apps/export/views.py, line 942 - E127 continuation line over-indented for visual indent
* corehq/apps/export/views.py, line 943 - E127 continuation line over-indented for visual indent
* corehq/apps/export/views.py, line 944 - E127 continuation line over-indented for visual indent
* corehq/apps/export/views.py, line 1158 - E501 line too long (125 > 115 characters)
* corehq/ex-submodules/couchexport/groupexports.py, line 51 - F841 local variable 'saved' is assigned to but never used
* corehq/couchapps/cached_export_by_export_instance_id/views/view/map.js, line 1 - Missing name in function declaration.

File: corehq/apps/export/views.py [hunk 1]
```diff
@@ -1115,10 +1154,13 @@ def fmt_export_data(self, export):
             emailed_exports = self.get_formatted_emailed_exports(export)
         else:
             # New export
-            emailed_exports = [export.daily_saved_export_metadata()]  # Have to wrap as array for legacy reasons
-
+            if export.is_daily_saved_export:
+                emailed_exports = [self._get_daily_saved_export_metadata(export)]  # Have to wrap as array for legacy reasons
```
Comment (dimagimon) [inline corehq/apps/export/views.py]: E501 line too long (125 > 115 characters)

File: corehq/couchapps/cached_export_by_export_instance_id/views/view/map.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+function(doc) {
```
Comment (dimagimon) [inline corehq/couchapps/cached_export_by_export_instance_id/views/view/map.js]: Missing name in function declaration.

**Comment by benrudolph** [comment]
looking good, i think i was confused a bit before seeing all the commits so some of my comments are irrelevant. the only thing i'm curious about is the `CachedExport`, maybe we can sync on it tomorrow? i'm imagining that there's a 1-to-1 relationship between `CachedExport` and `ExportInstance`? if it's not 1-to-1 i'd argue it probably could be. i saw your comment about separating presentation out, but isn't the `ExportInstance` the presentation and `ExportDataSchema` the model? 

 also one note about attachment names. it'd be nice if the attachment names meant something. we could do `<export_instance_id>_<last_updated>.<format>`. this way it's easy to find old exports and clean up the ones that aren't the latest with some sort of task/cron job

**Comment by NoahCarnahan** [comment]
Ben and I offlined about this. Seems like some things in the report module might use the `SavedBasicExport` (renamed to `CachedExport`) model, if they do we should probably keep `CachedExport` separate from `ExportInstance`. I'll do some digging here. The code also currently assumes there could be multiple `SavedBasicExport`s (renamed to `CachedExport`) for each saved export. I think this is because multiple async tasks could create them at once. If we do incorporate all the `CachedExport` fields in `ExportInstance` though, we can probably handle the document update conflict by just overwriting with the latest.
We also talked about the attachment names and decided it isn't necessary to give them specific names because the attachment gets overwritten when you update a daily saved export.

**Comment by NoahCarnahan** [comment]
I'm fairly certain there aren't any other models or parts of the code that use `SavedBasicExport`, so I'm going to merge `CachedExport` with `ExportInstance`.

**Comment by benrudolph** [comment]
sounds good :skull:

File: corehq/apps/export/export.py [hunk 3]
```diff
@@ -182,3 +186,40 @@ def _get_base_query(export_instance):
...
+    Rebuild the given daily saved ExportInstance
+    """
+    if _should_not_rebuild_export(export_instance, last_access_cutoff):
+        return
+
+    file = get_export_file([export_instance], filters or [])
+    with file as payload:
+        _save_export_payload(export_instance, payload)
+
+def _should_not_rebuild_export(export, last_access_cutoff):
```
Comment (dimagimon) [inline corehq/apps/export/export.py]: E302 expected 2 blank lines, found 1

File: corehq/apps/reports/urls.py [hunk 4]
```diff
@@ -85,6 +85,7 @@
     url(r"^export/bulk/download/$", "export_default_or_custom_data", name="export_bulk_download", kwargs=dict(bulk_export=True)),
     # saved
     url(r"^export/saved/download/(?P<export_id>[\w\-]+)/$", "hq_download_saved_export", name="hq_download_saved_export"),
+    url(r"^export/new/saved/download/(?P<export_instance_id>[\w\-]+)/$", "hq_download_new_saved_export", name="hq_download_new_saved_export"),
```
Comment (dimagimon) [inline corehq/apps/reports/urls.py]: E501 line too long (142 > 115 characters)

File: corehq/ex-submodules/couchexport/tests/test_saved.py [hunk 5]
```diff
@@ -1,11 +1,36 @@
 # coding=utf-8
 import datetime
+import uuid
```
Comment (dimagimon) [inline corehq/ex-submodules/couchexport/tests/test_saved.py]: F401 'uuid' imported but unused

**Comment by NoahCarnahan** [comment]
Just push some commits:
- Merged `CachedExport` into `ExportInstance`
- fixed a bug
- simplified some of the interface
- created a view for downloading daily saved `ExportInstance`

Daily saved exports are almost done, just have one or more things to do.

**Comment by benrudolph** [comment]
(let me know if you want this merged now or if you want to finish your one or two things)

File: corehq/apps/export/tasks.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 from celery.task import task
 
-from corehq.apps.export.export import get_export_file
+from corehq.apps.export.export import get_export_file, rebuild_export
+from corehq.apps.export.models.new import get_properly_wrapped_export_instance
```
Comment (dimagimon) [inline corehq/apps/export/tasks.py]: F401 'get_properly_wrapped_export_instance' imported but unused

File: corehq/apps/export/views.py [hunk 7]
```diff
@@ -1036,8 +1105,18 @@ def get_emailed_indexes(self, email_group):
         ExportConfiguration list"""
         raise NotImplementedError("must implement get_emailed_indexes")
 
+    def update_emailed_es_export_data(self, in_data):
+        from corehq.apps.export.tasks import rebuild_export_task
+        export_instance_id = in_data['export']['id']
+        export_instance = get_properly_wrapped_export_instance(export_instance_id)
```
Comment (dimagimon) [inline corehq/apps/export/views.py]: F841 local variable 'export_instance' is assigned to but never used

File: corehq/apps/reports/urls.py (hunk 4 above)
Comment (dimagimon) [inline corehq/apps/reports/urls.py]: E501 line too long (142 > 115 characters)

**Comment by NoahCarnahan** [comment]
Yeah merge this now, I'll put the next changes in another PR.

**Comment by benrudolph** [comment]
:+1: just waiting on build

File: corehq/apps/reports/urls.py (hunk 4 above)
Comment (dimagimon) [inline corehq/apps/reports/urls.py]: E501 line too long (142 > 115 characters)

File: corehq/couchapps/export_instances_by_is_daily_saved/views/view/map.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
+function(doc) {
```
Comment (dimagimon) [inline corehq/couchapps/export_instances_by_is_daily_saved/views/view/map.js]: Missing name in function declaration.

**Comment by NoahCarnahan** [comment]
:white_check_mark:

**Comment by benrudolph** [comment]
:dancer: :dancers: :dancer:

**Comment by czue** [comment]
wow! :fireworks: 

this is a pretty big change right? confirming you considered and decided no QA needed?

**Comment by NoahCarnahan** [comment]
It's still behind a feature flag, and there are still a few things left to implement before anyone should enable that feature flag, so not a huge change yet, but will be eventually.
I think the plan is:
- finish implementing all the exports changes behind flag
- have QA performed by enabling flag
- migrate existing exports and remove flag

**Pull request merged**